### DISCUSSION TITLE
[FEATURE REQUEST] add autorestart parameters to salt-minion service

### DISCUSSION THREAD

**Discussion by ewhitesides** [saltstack/salt#67635] [Feature Requests]
**Is your feature request related to a problem? Please describe.**
certain situations can come up, especially when updating a server, where the salt-minion service dies and doesn't come back up.  this can be a major inconvenience when working with several hundred servers.

**Describe the solution you'd like**
i found a lot of success defining a state that manages /etc/systemd/system/salt-minion.service.d/local.conf with the contents:
[Service]
Restart=on-failure
RestartSec=15
StartLimitBurst=3
StartLimitInterval=300

so if possible, could these (or similar) options be baked into the install as default?

**Describe alternatives you've considered**
tried experimenting with updating systems through [yumpkg.install](https://docs.saltproject.io/en/3000/ref/modules/all/salt.modules.yumpkg.html#salt.modules.yumpkg.install).  according to the documentation the module is supposed to isolate the process.

**Comment by sagetherage** [comment]
@ScriptAutomate only asking you to triage this ticket, please :-)

**Comment by sagetherage** [comment]
@ewhitesides 
Thank you for the feature request, we will review PRs if you can submit :)

**Comment by ewhitesides** [comment]
I'm not experienced enough to do a pull request.... I just suggested what I think the setting should be above.

**Comment by bryceml** [comment]
I'd like more people including some of the core team to weigh in on this before we do it, seems like it's likely a good change to make.

https://github.com/saltstack/salt-pack-py3/blob/develop/file_roots/pkg/salt/master/rhel8/sources/salt-minion.service is the current copy used for rhel8 for example.

I wonder if this should be added to the salt-master service, or other services as well.

I also see things like this: https://github.com/systemd/systemd/issues/16644  We should see which versions on which OSes require which syntax (in the Unit section vs the Service) and make sure we update them correctly for their given systemd version.

We also need to update the relevant files in https://gitlab.com/saltstack/open/salt-pkg at the same time.

**Comment by sagetherage** [comment]
@bryceml I will add it to the agenda for the core team to briefly review in the weekly Friday Planning meeting this week.

**Comment by bryceml** [comment]
We had a meeting, It sounded like the core team was hesitant to add this by default since different people may want different values depending on their situation, and some may expect it to not restart by default so errors don't keep happening or causing extra load.  

I think we said that we would bring this up in the community open hour on Thursdays to get more opinions.

**Comment by ewhitesides** [comment]
i just wanted to report back that this does not work well if you have a salt state that a) sets a salt-minion conf file and b) sets the salt-minion service to restart if the conf file changes.

in conjunction with restart settings (original post), you can end up in a loop where the state is trying to restart the service, and systemd is trying to restart the service.

my personal solution was to remove the restarts from the salt state, and define the conf files in terraform startup scripts before the service is started.  

that way the salt-minion shouldn't have a need to restart for a conf file change, and it will just restart if an unexpected disruption happens.

this all stems from that new module run syntax conf.... so hopefully this issue will be eliminated when it becomes the default in version 3005.

**Comment by waynew** [comment]
Well, FWIW we've realized that removing a module.run syntax is a horrible idea because it would break everything. But the different approaches are also unambiguous so we can inspect the format and decide whether you're trying to use old-style or new-style on a case-by-case basis, so presumably that will also solve the problem.

**Comment by xamindar** [comment]
I was just pointed to this. I can't believe this is still an issue. Makes updating salt environments a complete headache as the minion service dies on package update and stays dead. I would think this takes priority over other issues it might cause.

**Comment by ewhitesides** [comment]
@xamindar my solution was to first set up auto restart settings for the service (using a salt state). for example on Enterprise linux systems:
/etc/systemd/system/salt-minion.service.d/local.conf
[Service]
Restart=on-failure
RestartSec=15
StartLimitBurst=3
StartLimitInterval=300

windows machines seem to autorestart automatically without extra setttings. 

I then built a shell script which handles updating windows/linux minions from the master.
